clients/src/main/java/org/apache/kafka/common/errors/KafkaStorageException.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.errors;
+
+/**
+ * Miscellaneous disk-related IOException occurred when handling a request.
+ */
+public class KafkaStorageException extends ApiException {
```

Comment:
Hmm, if that is the case, should the storage exception be converted to LeaderNotAvailableException when it is thrown for requests such as produce/fetch/listOffsets? Also, in that case, does that mean the broker needs to hold on the response until the controller moves leadership? Otherwise broker is essentially giving up leadership by itself without get the confirmation from the controller.

Another thing is that we need to make sure the error code returned matches what is reflected in the metadata response.

I am not sure what is the best exception to return to the user in this case, but in general I would rather not reuse exceptions for different scenarios.